src/block/reader.rs

```diff
@@ -382,6 +384,154 @@ impl<R: ChunksReader> SequentialBlockDecompressor<R> {
+        let unbuffered = self.create_unbuffered_reader_at_position(position)?;
+        Ok(BufReader::new(unbuffered))
+    }
+}
+
+impl<T> MultiReadByteSource for std::io::Cursor<T> where T: AsRef<[u8]> + Send + Sync + Copy {
+    type Read = Self;
+    type BufferedRead = Self;
+
+    fn create_unbuffered_reader_at_position(&self, position: u64) -> Result<Self::Read> {
+        self.create_buffered_reader_at_position(position)
+    }
+
+    fn create_buffered_reader_at_position(&self, position: u64) -> Result<Self::BufferedRead> {
+        let mut clone = (*self).clone();
+        clone.seek(SeekFrom::Start(position))?;
+        Ok(clone)
+    }
+}
+
+// note: this is just pseudocode, it will be split up into a lot of separate pieces to allow for customization
+pub fn read_all_blocks_fully_parallel<PixelBlock>(
+    byte_source: impl MultiReadByteSource,
+    pixels_from_block: fn(UncompressedBlock) -> Result<PixelBlock>,
+    mut insert_block: impl FnMut(PixelBlock) -> UnitResult,
+    pedantic: bool
+) -> Result<()>
+    where PixelBlock: 'static + Send
+{
+    //println!("reading all blocks fully parallel...");
+
+    // TODO do not create a buffer for in-memory vectors
+    let mut read_buffered = PeekRead::new(Tracking::new(
+        byte_source.create_buffered_reader_at_position(0)?
+    ));
+
+    let meta = MetaData::read_from_buffered(&mut read_buffered, pedantic)?;
+    let offset_tables = MetaData::read_offset_tables(&mut read_buffered, &meta.headers)?;
+
+    {
+        if pedantic {
+            validate_offset_tables(
+                meta.headers.as_slice(), &offset_tables,
+                read_buffered.byte_position()
+            )?;
+        }
+
+        std::mem::drop(read_buffered); // TODO re-use when single threaded?
+    }
+
+    let mut sorted_chunk_offsets = offset_tables.iter().flatten().collect::<Vec<_>>();
+    sorted_chunk_offsets.sort_unstable();
+    //println!("processed offset tables");
+
+    let meta = Arc::new(meta);
+
+    let thread_count = 32;
+    let mut thread_pool = scoped_threadpool::Pool::new(thread_count)
+        // .thread_name("OpenEXR Block Decompressor".to_string())
+        // todo no more threads than remaining block count (self.len())
+        //.build();
+    ;
+
+    let (sender, receiver) = ::flume::unbounded(); // TODO bounded? bounded to threadpool size?
+    let sync_byte_source = Arc::new(byte_source); // TODO borrow instead?
+
+    //println!("starting thread pool work...");
+
+    // must be scoped because we want the multi byte source to contain byte slices, which have a lifetime
+    let result = thread_pool.scoped(|scope|{
+        for &chunk_location in sorted_chunk_offsets {
+            //println!("adding job for chunk at byte {}...", chunk_location);
+
+            let meta = meta.clone();
+            let sync_byte_source = sync_byte_source.clone();
+            let sender = sender.clone();
```

Comment:
we might get away without any `Arc`s because we use a scoped thread pool, which means we can borrow for the whole scope